src/xas_tdp_kernel.F

```diff
@@ -0,0 +1,1533 @@
+
+      !clean-up
+      CALL dbcsr_release(work_mat)
+      CALL dbcsr_deallocate_matrix_set(lhs_int)
+
+   END SUBROUTINE coulomb
+
+! **************************************************************************************************
+!> \brief Create the matrix containing the XC kenrel in the spin-conserving open-shell case:
+!>        (aI_sigma|fxc|J_tau b) ~= (aI_sigma|P) (P|Q)^-1 (Q|fxc|R) (R|S)^-1 (S|J_tau b)
+!> \param xc_ker the kernel matrix
+!> \param contr1_int_PQ the once contracted RI integrals, with inverse coulomb: (aI_sigma|P) (P|Q)^-1
+!> \param dist inherited dbcsr dist
+!> \param blk_size inherited block sizes
+!> \param donor_state ...
+!> \param xas_tdp_env ...
+!> \param xas_tdp_control ...
+!> \param qs_env ...
+!> note Prior to calling this function, the (Q|fxc|R) integral must be brodcasted to all procs
+! **************************************************************************************************
+   SUBROUTINE sc_os_xc(xc_ker, contr1_int_PQ, dist, blk_size, donor_state, xas_tdp_env, &
+                       xas_tdp_control, qs_env)
+
+      TYPE(dbcsr_type), INTENT(INOUT)                    :: xc_ker
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER          :: contr1_int_PQ
+      TYPE(dbcsr_distribution_type), POINTER             :: dist
+      INTEGER, DIMENSION(:), POINTER                     :: blk_size
+      TYPE(donor_state_type), POINTER                    :: donor_state
+      TYPE(xas_tdp_env_type), POINTER                    :: xas_tdp_env
+      TYPE(xas_tdp_control_type), POINTER                :: xas_tdp_control
+      TYPE(qs_environment_type), POINTER                 :: qs_env
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'sc_os_xc', routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: ndo_mo, ri_atom
+      LOGICAL                                            :: quadrants(3)
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER          :: lhs_int, rhs_int
+      TYPE(dbcsr_type)                                   :: work_mat
+
+      NULLIFY (lhs_int, rhs_int)
+
+      ! Initialization
+      ndo_mo = donor_state%ndo_mo
+      ri_atom = donor_state%at_index
+      !normal type work matrix such that distribution of all spin quadrants match
+      CALL dbcsr_create(work_mat, name="WORK", matrix_type="N", dist=dist, &
+                        row_blk_size=blk_size, col_blk_size=blk_size)
+
+      rhs_int => contr1_int_PQ ! contains [ (aI|P)*(P|Q)^-1 ]^T
+      ALLOCATE (lhs_int(SIZE(contr1_int_PQ))) ! will contain (aI|P)*(P|Q)^-1 * (Q|fxc|R)
+
+      ! Case study: UKS or ROKS ?
+      IF (xas_tdp_control%do_uks) THEN
+
+         ! In the case of UKS, donor MOs might be different for different spins. Moreover, the
+         ! fxc itself might change since fxc = fxc_sigma,tau
+         ! => Carfully treat each spin-quadrant separately
+
+         ! alpha-alpha spin quadrant (upper-lefet)
+         quadrants = .FALSE.; quadrants(1) = .TRUE.
```

Comment:
wondering whether `quadrants = [.TRUE., .FALSE., .FALSE.]` might not be a bit clearer